### ISSUE TITLE
Enhancement: Use steady state solution to restart another simulation

### ISSUE THREAD

**Issue by abuntrock** [OpenModelica/OpenModelica#11787]
### Description

When simulating large hydraulic models our BRSL users often times want them to start in steady state. This could be made easier when this steady state solution could be deduced by a previous simulation. The steady state solution of this previous simulation could then be used to initialize the model in this same steady state. We know that this is a complicated issue but we nevertheless wanted to show that we are interested in this topic.

This might be a duplicate of the first point in this stackoverflow post
https://stackoverflow.com/questions/42673420/global-measure-to-test-when-steady-state-of-system-is-reached
which made it to this ticket in the old ticketing system 
https://trac.openmodelica.org/OpenModelica/ticket/4301
and then into GitHub here
https://github.com/OpenModelica/OpenModelica/issues/4301.

If you think this is a duplicate, feel free to close this issue. 

### Expected Behavior

I can simulate a model until it reaches a steady state. I can then restart the simulation initialized in this steady state.

### Version and OS

 - OpenModelica Version: 1.22.1
 - OS: Windows 10, 64 bit

Thanks for considering this.
Aaron Buntrock - Bosch Rexroth

**Comment by casella** [comment]
@abuntrock this feature is described in the [User's Guide](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/solving.html#importing-initial-values-from-previous-simulations). Please reopen if doesn't work as you need.

**Comment by abuntrock** [comment]
@casella Thanks for pointing that out. I haven't looked there for this problem, excuse me. 
The scenario with `iim = "none" ` is basically what I'm looking for.
However, I have two remarks:

1. In my case i Just cannot simply paste the example code from the UserGuide and run `M2` (or `M3`) directly after `M`. I need to copy the inital.mat file created by `M` to the working folder of `M2` or I need to change the `iif` setting in `M2` to point to the working folder of `M`. Maybe I am doing something wrong or this is considered obvious but I would think it could be useful to add this to the UserGuide.

**Comment by casella** [comment]
>     1. In my case i Just cannot simply paste the example code from the UserGuide and run `M2` (or `M3`) directly after `M`. I need to copy the inital.mat file created by `M` to the working folder of `M2` or I need to change the `iif` setting in `M2` to point to the working folder of `M`. Maybe I am doing something wrong or this is considered obvious but I would think it could be useful to add this to the UserGuide. 

You are right, I guess I wrote this part of the Guide before we changed the organization of the working directory. In order to make this work, we need to use `"../initial.mat"` as filename, so that it is placed in the root working directory and is accessible by all models from the package. I just opened PR #11797 to fix this in the User Guide.

> This copying is also a stumbling block from my point of view because in our field one typically wants to use simulation results from a model to restart the same model. To do this one has to make sure, to change the name of the output file in between the runs.

Yes, currently you have to do it manually with the operating system's file manager. We could add one more button Continue Simulation beside Re-Simulate and Re-Simulate Setup, which automatically sets the initial time to the final time of the previously saved simulation, manages the result files and sets the -iim and -iif parameters as needed. The question is, when you continue the simulation, do you want to overwrite the previous result file or keep it?

**Comment by casella** [comment]
> 2. I am not currently not able to apply this feature to a complicated BRSL model (`Rexroth_BRSL.Examples.Systems.SingleAxis` from the BRSL3.0, which should be available to you). I tried the following workflow.
> 
> 
>     * Simulate the model once.
> 
>     * Save the resulting .mat file in another location.
> 
>     * Use the .mat file as the `iif` input and set `iim = "none"`.
> 
>     * Simulate the model again.
>       From my understanding this should always work, since the values at time zero from the first simulation are just used again to start the second simulation instead of solving the initialization problem again. However, I am getting a lot of warnings like
>       `unable to import real variable velCommand.derivative.x from given file` and the simulation aborts because of a division by zero at time=0. Maybe the reason is that all of our subcomponents are protected and are thus not written to the .mat file?

Absolutely. If you want this feature to work, you need to save all the variables, including the protected ones.

**Comment by abuntrock** [comment]
> Yes, currently you have to do it manually with the operating system's file manager. We could add one more button Continue Simulation beside Re-Simulate and Re-Simulate Setup, which automatically sets the initial time to the final time of the previously saved simulation, manages the result files and sets the -iim and -iif parameters as needed. The question is, when you continue the simulation, do you want to overwrite the previous result file or keep it?

This sounds like a helpful idea. In our use-case one typically wants to keep the previous result file. There is normally one starting simulation to find good initial values and then there are multiple simulations based on those good initial values. But there are probably other use-cases where one wants to overwrite the previous result file.

**Comment by abuntrock** [comment]
> Absolutely. If you want this feature to work, you need to save all the variables, including the protected ones.

Thanks for the clarification. In our BRSL classes we need some protection annotations in combination with encryption because of know-how protection issues. But then we have to live with the fact that those two things are incompatible right now.

**Comment by casella** [comment]
> Thanks for the clarification. In our BRSL classes we need some protection annotations in combination with encryption because of know-how protection issues. But then we have to live with the fact that those two things are incompatible right now.

There is a fundamental conflict between equation-based modelling and encryption. Issues are going to pop up every now and then...

**Comment by casella** [comment]
> This sounds like a helpful idea. In our use-case one typically wants to keep the previous result file. There is normally one starting simulation to find good initial values and then there are multiple simulations based on those good initial values. But there are probably other use-cases where one wants to overwrite the previous result file.

> OK, so we could have a "Continue Simulation" button alongside the Re-Simulate button, which brings up a similar dialog with StartTime already set to the StopTime of the first simulation and with the appropriate flags set. 

We have a choice though. One option is to always start from the end of the latest simulation; this would allow to simulate the first 1000 s, then the next 1000 s, then the next 1000 s, and so on. The other one is to always restart from the end of the latest simulation  (not continued simulation), which is what you wrote you need above.

We could have a checkbox to distinguish the two cases. Not sure what is the best text to express that, maybe "Always restart from the end of the first simulation"?

**Comment by achniouel** [comment]
I'm facing a similar issue, but my situation involves two models: one is stationary, and the other is transient. My goal is to transfer the variable values from the stationary model to the transient one. However, when I attempt this, I encounter an error message stating: "unable to import boolean parameter Entree_RRI.use_C_in from given file," among others. I've tried following the above-mentioned workflow without success.

**Comment by bilderbuchi** [comment]
> We could add one more button Continue Simulation beside Re-Simulate and Re-Simulate Setup, which automatically sets the initial time to the final time of the previously saved simulation, manages the result files and sets the -iim and -iif parameters as needed. The question is, when you continue the simulation, do you want to overwrite the previous result file or keep it?

This sounds like a very useful addition to us. 
As to your question, ideally the result file can be appended/concatenated (it's fine if this is one of several options), because our use case is what you describe above
> this would allow to simulate the first 1000 s, then the next 1000 s, then the next 1000 s, and so on. 

and we would like to then see in the plots the results for the whole 3000 s.

**Comment by j-emils** [comment]
I have a very similar need. I would like to run a simulation (any time duration) and then change a replaceable component and restart the simulation getting the initial conditions and (possibly) the start-time from the previous simulation.
